### PR TITLE
Feat: implement pipeline APIs

### PR DESCRIPTION
**Pull request by chivalryq**
Signed-off-by: Qiaozp <qiaozhongpei.qzp@alibaba-inc.com>
Co-authored-by: FogDong [dongtianxin.tx@alibaba-inc.com](mailto:dongtianxin.tx@alibaba-inc.com)

### Description of your changes

Implement pipeline APIs

Fixes #

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/apiserver/interfaces/api/pipeline.go [hunk 1]
```diff
@@ -276,43 +320,39 @@ func (p *pipelineAPIInterface) updatePipeline(req *restful.Request, res *restful
...
-func (p *pipelineAPIInterface) deletePipeline(req *restful.Request, res *restful.Response) {
+func (n *projectAPIInterface) deletePipeline(req *restful.Request, res *restful.Response) {
 	pipeline := req.Request.Context().Value(&apis.CtxKeyPipeline).(apis.PipelineBase)
-	err := p.PipelineService.DeletePipeline(req.Request.Context(), pipeline)
+	err := n.PipelineService.DeletePipeline(req.Request.Context(), pipeline)
 	if err != nil {
 		log.Logger.Errorf("delete pipeline failure %s", err.Error())
 		bcode.ReturnError(req, res, err)
 		return
 	}
```
Comment (barnettZQG) [inline pkg/apiserver/interfaces/api/pipeline.go:330]: Please return an empty struct. `apis.EmptyResponse{}`

File: pkg/apiserver/domain/model/pipeline.go [hunk 2]
```diff
@@ -28,6 +31,39 @@ type Value struct {
...
+	return "pipeline"
+}
+
+// Index return custom index
+func (p Pipeline) Index() map[string]string {
+	var index = make(map[string]string)
+	if p.Project != "" {
+		index["project"] = p.Project
+	}
+	return index
```
Comment (barnettZQG) [inline pkg/apiserver/domain/model/pipeline.go:64]: Make the name an index key.

File: pkg/apiserver/domain/service/pipeline.go [hunk 3]
```diff
@@ -110,43 +124,92 @@ func NewContextService() ContextService {
...
-	for _, wf := range wfs.Items {
-		if fuzzyMatch(wf, req.Query) {
+	for _, _p := range pipelines {
+		pipeline := _p.(*model.Pipeline)
+		if !pkgutils.StringsContain(availableProjectNames, pipeline.Project) {
+			continue
+		}
+		if fuzzyMatch(pipeline, req.Query) {
+			base := pipeline2PipelineBase(pipeline, projectMap[pipeline.Project])
+			info, err := p.getPipelineInfo(ctx, pipeline, projectMap[pipeline.Project].Namespace)
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:199]: Add a request param to control whether to query info. The default is True.

File: pkg/apiserver/domain/service/pipeline.go [hunk 4]
```diff
@@ -19,74 +19,88 @@ package service
...
+	querytypes "github.com/oam-dev/kubevela/pkg/velaql/providers/query/types"
 )
 
 const (
-	labelContextName = "context.velaux.oam.dev/name"
+	annotationDescription = "pipeline.oam.dev/description"
+	annotationAlias       = "pipeline.oam.dev/alias"
+	labelProject          = "pipeline.oam.dev/project"
+	labelContext          = "pipeline.oam.dev/context"
+	labelPipeline         = "pipeline.oam.dev/pipeline"
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:53]:
```suggestion
	labelPipeline         = "pipeline.oam.dev/name"
```

File: pkg/apiserver/domain/service/pipeline.go [hunk 5]
```diff
@@ -110,43 +124,92 @@ func NewContextService() ContextService {
...
+
+	}
+	if len(availableProjectNames) == 0 || len(namespaces) == 0 {
+		return &apis.ListPipelineResponse{}, nil
+	}
+	if len(namespaces) == 1 {
+		nsOption = append(nsOption, client.InNamespace(projectNamespace[req.Projects[0]]))
+	}
+	pp := model.Pipeline{}
+	pipelines, err := p.Store.List(ctx, &pp, nil)
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:187]: You should list by filter params.

File: pkg/apiserver/domain/service/pipeline.go [hunk 6]
```diff
@@ -155,45 +218,89 @@ func (p pipelineServiceImpl) ListPipelines(ctx context.Context, req apis.ListPip
...
+		}
 		return nil, err
 	}
-	wf.WorkflowSpec = req.Spec
-	wf.SetLabels(map[string]string{
-		labelDescription: req.Description,
-		labelAlias:       req.Alias})
-	if err := p.KubeClient.Update(ctx, &wf); err != nil {
+
+	pipeline.Spec = req.Spec
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:266]: Add a test to cover updating the step mode and sub-step mode.

File: pkg/apiserver/domain/service/pipeline.go [hunk 7]
```diff
@@ -110,43 +124,92 @@ func NewContextService() ContextService {
...
-		labelDescription: req.Description,
-		labelAlias:       req.Alias})
-	if err := p.KubeClient.Create(ctx, &wf); err != nil {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	pipeline := &model.Pipeline{
+		Name:        req.Name,
+		Description: req.Description,
+		Alias:       req.Alias,
+		Project:     project.Name,
+		Spec:        req.Spec,
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:133]: Check the `req.Spec`. Such as:

1. The steps can not  be empty.
2. The mode is DAG or StepByStep.

File: pkg/apiserver/domain/service/pipeline.go [hunk 8]
```diff
@@ -155,45 +218,89 @@ func (p pipelineServiceImpl) ListPipelines(ctx context.Context, req apis.ListPip
...
-	wf := v1alpha1.Workflow{}
-	if err := p.KubeClient.Get(ctx, client.ObjectKey{Name: pl.Name, Namespace: nsForProj(pl.Project)}, &wf); err != nil {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	pipeline := &model.Pipeline{
+		Name:    pl.Name,
+		Project: project.Name,
+	}
+	if err := p.Store.Get(ctx, pipeline); err != nil {
+		if errors.Is(err, datastore.ErrRecordNotExist) {
+			return nil
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:285]: Can not ignore this error. The API should return 404.

File: pkg/apiserver/domain/service/pipeline.go [hunk 9]
```diff
@@ -155,45 +218,89 @@ func (p pipelineServiceImpl) ListPipelines(ctx context.Context, req apis.ListPip
...
+	if err := p.ContextService.DeleteAllContexts(ctx, pl.Project.Name, pl.Name); err != nil {
+		log.Logger.Errorf("delete pipeline all context failure: %s", err.Error())
 		return err
 	}
-	return p.KubeClient.Delete(ctx, &wf)
+	if err := p.PipelineRunService.CleanPipelineRuns(ctx, pl); err != nil {
+		log.Logger.Errorf("delete pipeline all pipeline-runs failure: %s", err.Error())
+		return err
+	}
+	return nil
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:298]: Did not delete the Pipeline?

File: pkg/apiserver/domain/service/pipeline.go [hunk 10]
```diff
@@ -155,45 +218,89 @@ func (p pipelineServiceImpl) ListPipelines(ctx context.Context, req apis.ListPip
...
 }
 
 // StopPipelineRun will stop a pipelineRun
 func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun apis.PipelineRunBase) error {
-	run, err := p.checkRecordRunning(ctx, pipelineRun)
+	run, err := p.checkRecordNotFinished(ctx, pipelineRun)
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:303]:
```suggestion
	run, err := p.checkRunNotFinished(ctx, pipelineRun)
```

File: pkg/apiserver/domain/service/pipeline.go [hunk 11]
```diff
@@ -203,53 +310,542 @@ func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun
...
-	// todo get context and set to workflowRun
+func (p pipelineServiceImpl) RunPipeline(ctx context.Context, pipeline apis.PipelineBase, req apis.RunPipelineRequest) (*apis.PipelineRun, error) {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
 	run := v1alpha1.WorkflowRun{}
 	version := utils.GenerateVersion("")
 	name := fmt.Sprintf("%s-%s", pipeline.Name, version)
 	run.Name = name
-	run.Namespace = fmt.Sprintf("%s-project", pipeline.Project)
+	run.Namespace = project.GetNamespace()
 	run.Spec.WorkflowRef = pipeline.Name
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:628]: Assign the value for the Steps and not WorkflowRef

File: pkg/apiserver/domain/service/pipeline.go [hunk 12]
```diff
@@ -329,99 +938,161 @@ func (c contextServiceImpl) DeleteContext(ctx context.Context, projectName, pipe
...
+					Name:  project.Name,
+					Alias: project.Alias,
+				},
 				PipelineRunName: run.Name,
 			},
+			Spec: run.Spec,
 		},
 		Status: run.Status,
 	}
+	ctxName := run.GetLabels()[labelContext]
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:1006]: The result of GetLabels() may be nil.

File: pkg/utils/k8s.go [hunk 13]
```diff
@@ -199,3 +211,124 @@ func IsClusterScope(gvk schema.GroupVersionKind, mapper meta.RESTMapper) (bool,
...
+			}
+			// 48h
+			dur, _ := time.ParseDuration("48h")
+			tail := stern.NewTail(p.Namespace, p.Pod, p.Container, template, &stern.TailOptions{
+				Timestamps:   true,
+				SinceSeconds: int64(dur.Seconds()),
+				Exclude:      nil,
+				Include:      nil,
+				Namespace:    false,
+				TailLines:    nil, // default for all logs
```
Comment (barnettZQG) [inline pkg/utils/k8s.go:289]: Must set the max lines.

File: pkg/apiserver/domain/service/pipeline.go (hunk 12 above)
Comment (chivalryq) [reply pkg/apiserver/domain/service/pipeline.go:1006]: Why？

Comment (barnettZQG) [reply pkg/apiserver/domain/service/pipeline.go:1006]: If users create the WorkflowRun by CLI.

File: pkg/apiserver/infrastructure/clients/kubeclient.go [hunk 14]
```diff
@@ -77,7 +80,26 @@ func GetKubeClient() (client.Client, error) {
...
 	})
-	return kubeClient, err
+	if err != nil {
+		return nil, err
+	}
+	err = v1alpha1.AddToScheme(common.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	kubeClient, err = multicluster.Initialize(kubeConfig, false)
```
Comment (wonderflow) [inline pkg/apiserver/infrastructure/clients/kubeclient.go:90]: we have aleady created one in L80 above.

File: pkg/apiserver/interfaces/api/dto/v1/types.go [hunk 15]
```diff
@@ -1551,10 +1554,11 @@ type ListConfigDistributionResponse struct {
...
 type PipelineMeta struct {
-	Name        string `json:"name" validate:"checkname"`
-	Alias       string `json:"alias" validate:"checkalias" optional:"true"`
-	Project     string `json:"project"`
-	Description string `json:"description" optional:"true"`
+	Name        string    `json:"name" validate:"checkname"`
+	Alias       string    `json:"alias" validate:"checkalias" optional:"true"`
+	Project     NameAlias `json:"project"`
+	Description string    `json:"description" optional:"true"`
+	CreateTime  time.Time `json:"createTime"`
```
Comment (wonderflow) [inline pkg/apiserver/interfaces/api/dto/v1/types.go:1561]: optional?

File: pkg/apiserver/interfaces/api/dto/v1/types.go [hunk 16]
```diff
@@ -1629,9 +1633,8 @@ type GetPipelineResponse struct {
 
 // PipelineInfo is the info of pipeline
 type PipelineInfo struct {
-	RelatedApps   []ApplicationBase                  `json:"relatedApps"`
```
Comment (wonderflow) [inline pkg/apiserver/interfaces/api/dto/v1/types.go:1632]: why this field `relatedApps` removed?

File: pkg/apiserver/interfaces/api/pipeline.go [hunk 17]
```diff
@@ -49,168 +43,211 @@ const (
...
+		Param(ws.QueryParameter("detailed", "query pipelines with detail").DataType("bool").DefaultValue("true")).
+		Returns(200, "OK", apis.ListPipelineResponse{}).
+		Returns(400, "Bad Request", bcode.Bcode{}).
+		Writes(apis.ListPipelineResponse{}).Do(meta))
+
 	ws.Filter(authCheckFilter)
 	return ws
 }
 
+type pipelineAPIInterface struct {
```
Comment (wonderflow) [inline pkg/apiserver/interfaces/api/pipeline.go:233]:
```suggestion
type pipelineAPI struct {
```

This is an struct not interface.

File: pkg/apiserver/interfaces/api/pipeline.go [hunk 18]
```diff
@@ -49,168 +43,211 @@ const (
...
 	return &pipelineAPIInterface{}
 }
 
-func (p *pipelineAPIInterface) listPipelines(req *restful.Request, res *restful.Response) {
-	var projetNames []string
-	if req.QueryParameter("project") != "" {
-		projetNames = append(projetNames, req.QueryParameter("project"))
+func (n *pipelineAPIInterface) listPipelines(req *restful.Request, res *restful.Response) {
+	var projectNames []string
+	if req.QueryParameter("projectName") != "" {
```
Comment (wonderflow) [inline pkg/apiserver/interfaces/api/pipeline.go:244]: use a const variable for this keyword

File: pkg/utils/k8s.go [hunk 19]
```diff
@@ -199,3 +211,124 @@ func IsClusterScope(gvk schema.GroupVersionKind, mapper meta.RESTMapper) (bool,
 	isClusterScope := len(mappings) > 0 && mappings[0].Scope.Name() == meta.RESTScopeNameRoot
 	return isClusterScope, err
 }
+
+// GetPodsLogs get logs from pods
+func GetPodsLogs(ctx context.Context, config *rest.Config, containerName string, selectPods []*querytypes.PodBase, tmpl string, logC chan<- string, tailLines *int64) error {
```
Comment (wonderflow) [inline pkg/utils/k8s.go:216]: why don't re-use vela logs CLI code?

File: pkg/utils/k8s.go [hunk 20]
```diff
@@ -199,3 +211,124 @@ func IsClusterScope(gvk schema.GroupVersionKind, mapper meta.RESTMapper) (bool,
...
+				Exclude:      nil,
+				Include:      nil,
+				Namespace:    false,
+				TailLines:    tailLines, // default for all logs
+			})
+			tails[id] = tail
+
+			tail.Start(ctx, clientSet.CoreV1().Pods(p.Namespace), logC)
+		}
+	}()
```
Comment (wonderflow) [inline pkg/utils/k8s.go:295]: when will the go routine finished? Will it become a leak?

File: pkg/utils/k8s.go [hunk 21]
```diff
@@ -199,3 +211,124 @@ func IsClusterScope(gvk schema.GroupVersionKind, mapper meta.RESTMapper) (bool,
...
+	go func() {
+		for p := range removed {
+			id := p.GetID()
+			if tails[id] == nil {
+				continue
+			}
+			tails[id].Close()
+			delete(tails, id)
+		}
+	}()
```
Comment (wonderflow) [inline pkg/utils/k8s.go:306]: same above

File: pkg/utils/k8s.go (hunk 19 above)
Comment (wonderflow) [reply pkg/utils/k8s.go:216]: sorry, I see you've re-used the func.

File: pkg/apiserver/domain/model/pipeline.go [hunk 22]
```diff
@@ -16,12 +16,53 @@ limitations under the License.
...
 
-import "fmt"
+import (
+	"fmt"
+
+	"github.com/kubevela/workflow/api/v1alpha1"
+)
 
 func init() {
 	RegisterModel(&PipelineContext{})
```
Comment (barnettZQG) [inline pkg/apiserver/domain/model/pipeline.go:26]: Miss the Pipeline?

File: pkg/apiserver/domain/service/pipeline.go [hunk 23]
```diff
@@ -110,90 +122,195 @@ func NewContextService() ContextService {
...
+		Name:    pl.Name,
+		Project: project.Name,
+	}
+	if err := p.Store.Get(ctx, pipeline); err != nil {
+		if errors.Is(err, datastore.ErrRecordNotExist) {
+			return bcode.ErrPipelineNotExist
+		}
+		return err
+	}
+	if err := p.Store.Delete(ctx, pipeline); err != nil {
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:297]: Delete the pipeline after the run deletion?

File: pkg/apiserver/domain/service/pipeline.go [hunk 24]
```diff
@@ -203,53 +320,544 @@ func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun
...
+			return apis.GetPipelineRunLogResponse{
+				StepBase: getStepBase(pipelineRun, step),
+				Log:      "",
+			}, nil
+		}
+		return apis.GetPipelineRunLogResponse{}, err
+	}
+	var logs string
+	switch {
+	case logConfig.Data:
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:463]: What means `logConfig.Data`?  Special implement for the job logs?

File: pkg/apiserver/domain/service/pipeline.go [hunk 25]
```diff
@@ -203,53 +320,544 @@ func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun
...
+	case logConfig.Source != nil:
+		if len(logConfig.Source.Resources) > 0 {
+			logs, err = getResourceLogs(ctx, p.KubeConfig, p.KubeClient, logConfig.Source.Resources, nil)
+			if err != nil {
+				return apis.GetPipelineRunLogResponse{}, err
+			}
+		}
+		if logConfig.Source.URL != "" {
+			var logsBuilder strings.Builder
+			readCloser, err := wfUtils.GetLogsFromURL(ctx, logConfig.Source.URL)
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/pipeline.go:480]: Limit body size?

File: pkg/apiserver/interfaces/api/dto/v1/types.go (hunk 16 above)
Comment (chivalryq) [reply pkg/apiserver/interfaces/api/dto/v1/types.go:1632]: @barnettZQG canceled this field.

File: pkg/apiserver/interfaces/api/pipeline.go (hunk 17 above)
Comment (chivalryq) [reply pkg/apiserver/interfaces/api/pipeline.go:235]: This `Interface` seems to mean "network interface" not "interface in golang". And this is the naming rules for all other similar struct.

File: pkg/apiserver/domain/service/pipeline.go (hunk 24 above)
Comment (chivalryq) [reply pkg/apiserver/domain/service/pipeline.go:463]: See log config definition here: https://github.com/chivalryq/kubevela/blob/75760521a460d881d206739fb1c33ccd5884f50c/pkg/stdlib/pkgs/util.cue
https://github.com/kubevela/kubevela/blob/75760521a460d881d206739fb1c33ccd5884f50c/pkg/stdlib/pkgs/util.cue

File: pkg/apiserver/interfaces/api/pipeline.go (hunk 17 above)
Comment (wonderflow) [reply pkg/apiserver/interfaces/api/pipeline.go:235]: This naming rule seems confusion, it's bad.

File: pkg/apiserver/domain/service/pipeline.go (hunk 23 above)
Comment (chivalryq) [reply pkg/apiserver/domain/service/pipeline.go]: change the order to run->context->pipeline

File: pkg/apiserver/interfaces/api/dto/v1/types.go (hunk 15 above)
Comment (chivalryq) [reply pkg/apiserver/interfaces/api/dto/v1/types.go:1561]: remove optional tag in returned values

File: pkg/apiserver/interfaces/api/pipeline.go (hunk 17 above)
Comment (chivalryq) [reply pkg/apiserver/interfaces/api/pipeline.go:231]: We can change the naming uniformly. How do you like that @barnettZQG

File: pkg/apiserver/domain/service/rbac.go [hunk 26]
```diff
@@ -234,6 +234,17 @@ var ResourceMaps = map[string]resourceMetadata{
 				pathName: "configName",
 			},
 			"provider": {},
+			"pipeline": {
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/rbac.go:237]: Set the default role and permissions for the Pipeline resource.

File: pkg/apiserver/domain/service/suite_test.go [hunk 27]
```diff
@@ -25,6 +25,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/kubevela/workflow/api/v1alpha1"
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/suite_test.go:28]: Sort the imports.

Comment (chivalryq) [reply pkg/apiserver/domain/service/suite_test.go:28]: this is already the right sort. kubevela/workflow is other package.

File: pkg/apiserver/domain/service/pipeline_test.go [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+					Name: "step-group",
+					Type: "step-group",
+				},
+			},
+		}
+
+		By("create pipeline with sub-steps")
+		pipeline, err := pipelineService.CreatePipeline(ctx, apisv1.CreatePipelineRequest{
+			Name: pipelineName,
+			Spec: v1alpha1.WorkflowSpec{
```
[the remark this answers is not in the document]
Comment (chivalryq) [reply pkg/apiserver/domain/service/pipeline_test.go:99]: Already test it here: 
https://github.com/kubevela/kubevela/blob/d9e8fd03426c811d741dff5567642f5d42f0ad3f/test/e2e-apiserver-test/pipeline_test.go L213

File: pkg/apiserver/domain/service/rbac.go (hunk 26 above)
Comment (chivalryq) [reply pkg/apiserver/domain/service/rbac.go:248]: done

File: references/cli/logs.go [hunk 29]
```diff
@@ -158,69 +130,11 @@ func (l *Args) printPodLogs(ctx context.Context, ioStreams util.IOStreams, selec
...
-			id := p.GetID()
-			if tails[id] == nil {
-				continue
-			}
-			tails[id].Close()
-			delete(tails, id)
-		}
-	}()
-
 	<-ctx.Done()
```
Comment (wonderflow) [inline references/cli/logs.go:138]: do we need to close logC channel here？

File: pkg/utils/k8s.go [hunk 30]
```diff
@@ -199,3 +211,125 @@ func IsClusterScope(gvk schema.GroupVersionKind, mapper meta.RESTMapper) (bool,
...
+
+	<-ctx.Done()
```
Comment (wonderflow) [inline pkg/utils/k8s.go:308]: dup

File: references/cli/logs.go (hunk 29 above)
Comment (wonderflow) [inline references/cli/logs.go:138]: dup

File: pkg/apiserver/domain/service/pipeline.go [hunk 31]
```diff
@@ -203,53 +330,625 @@ func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun
...
+		})
+	}
+	o.Values = values
+	return o
+}
+
+func getResourceLogs(ctx context.Context, config *rest.Config, cli client.Client, resources []wfTypes.Resource, filters []string) (string, error) {
+	var (
+		linesOfLogKept int64 = 1000
+		timeout              = 5 * time.Second
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/pipeline.go:604]:
```suggestion
		timeout              = 10 * time.Second
```

**Review by wonderflow — APPROVED** [review]

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**